core/src/main/scala/filodb.core/memstore/IndexMetadataStore.scala

```diff
@@ -86,6 +91,249 @@ trait IndexMetadataStore {
+        getSnapTime(snapFile.toString, fin) match {
+          case Success(snapTime)    =>  hour() - hour(snapTime) > maxRefreshHours
+          case Failure(_)           =>  true
+        }
+      }.get
+    } else
+      false
+
+/**
+   * The /<rootDirectory>/_<dataset>_<shard>_snap file contains the last time the index for the given dataset and shard
+   * was synched. Current implementation only supports storing Synced state and no state related information is stored
+   * in the snap file. The file contains the magic byte followed by 8 bytes for the long value for the timestamp
+   * starting with most to least significant bytes
+   *
+   * In case the snap file does not exists, magic is not as expected or the file is corrupt, RebuildIndex will be
+   * triggered
+   *
+   * @param datasetRef The dataset ref
+   * @param shard the shard id of index
+   * @return a tuple of state and the option of time at which the state was recorded
+   */
+
+  override def currentState(datasetRef: DatasetRef, shard: Int): (IndexState.Value, Option[Long]) = {
+    val snapFile = new File(snapFilePathFormat.format(datasetRef.dataset, shard))
+    if (!snapFile.exists()) {
+      logger.info("Snap file {} does not exist , triggering rebuild", snapFile.toString)
+      (IndexState.Empty, None)
+    } else {
+      Using(new FileInputStream(snapFile)) { fin =>
+        getSnapTime(snapFile.toString, fin) match {
+          case Success(snapTime)    =>
+                                        logger.info("Found lastSyncTime={} for dataset={} and shard={}",
+                                          snapTime, datasetRef.toString, shard)
+                                        (IndexState.Synced, Some(snapTime))
+          case Failure(e)           =>  throw e           // Failure to read the snap file in currentState is fatal
+        }
+      }.get
+    }
+  }
+
+  /**
+   *  Gets the snapTime from the snap file
+   *
+   * @param snapFile   the snapFile
+   * @param fin        The input stream object for the file
+   * @return           Long snap time if file is valid
+   */
+  private def getSnapTime(snapFile: String, fin: FileInputStream): Try[Long] = {
+        val magicByte = fin.read()
+        if (magicByte != FileSystemBasedIndexMetadataStore.snapFileV1Magic) {
+          // if first byte is not the expected magicByte then the file is not as expected and possibly garbled
+          logger.warn(" Snap {} exist and possibly corrupt, this will trigger a rebuild", snapFile.toString)
+          Failure(new IllegalStateException("Invalid magic bytes in snap file"))
+        } else {
+          // If the first byte is magicByte read just the next 8 bytes
+          val bytes = Array[Byte](0, 0, 0, 0, 0, 0, 0, 0)
+          if (fin.read(bytes) != 8) {
+            // Not able to read 8 bytes, the file is truncated and we will trigger a rebuild
+            logger.warn("Snap file {} with right magic bytes but less than 8 for time," +
```

Comment:
NIT: wondering if we should use logger.error here ? (since currentState throws exception, which means the initState could also throw exception, but I am not sure if that will lead to some fatal error)

Reply:
Changing the log level would not cause any issue. I just logged a warn since I am returning a failure anyways,